mlrun/api/db/sqldb/db.py

```diff
@@ -2438,59 +3294,125 @@ def store_feature_set(
-                ):
+                if self._re_tag_existing_object(session, cls, project, name, tag, uid):
                     return uid
 
-                db_feature_set = FeatureSet(project=project)
+                db_tagged_object = cls(project=project)
 
             self._update_db_record_from_object_dict(
-                db_feature_set, feature_set_dict, uid
+                db_tagged_object, tagged_object_dict, uid
             )
 
-            self._update_feature_set_spec(db_feature_set, feature_set_dict)
-            self._upsert(session, [db_feature_set])
+            if cls == FeatureSet:
+                self._update_feature_set_spec(db_tagged_object, tagged_object_dict)
+            self._upsert(session, [db_tagged_object])
             if tag:
-                self.tag_objects_v2(session, [db_feature_set], project, tag)
+                self.tag_objects_v2(session, [db_tagged_object], project, tag)
             return uid
 
         # Object with the given tag/uid doesn't exist
         # Check if this is a re-tag of existing object - search by uid only
-        if self._re_tag_existing_object(session, FeatureSet, project, name, tag, uid):
+        if self._re_tag_existing_object(session, cls, project, name, tag, uid):
             return uid
 
-        feature_set.metadata.tag = tag
-        return self.create_feature_set(session, project, feature_set, versioned)
+        tagged_object.metadata.tag = tag
+        return self._create_tagged_object(
+            session, project, cls, tagged_object, versioned
+        )
+
+    def _create_tagged_object(
+        self,
+        session,
+        project,
+        cls,
+        tagged_object: typing.Union[
+            mlrun.common.schemas.FeatureVector,
+            mlrun.common.schemas.FeatureSet,
+            dict,
+        ],
+        versioned=True,
+    ):
+        (uid, tag, tagged_object_dict,) = self._validate_and_enrich_record_for_creation(
+            session, tagged_object, cls, project, versioned
+        )
+
+        db_tagged_object = cls(project=project)
 
-    def _re_tag_existing_object(self, session, cls, project, name, tag, uid):
-        _, _, existing_object = self._get_record_by_name_tag_and_uid(
-            session, cls, project, name, None, uid
+        self._update_db_record_from_object_dict(
+            db_tagged_object, tagged_object_dict, uid
         )
+        if cls == FeatureSet:
+            self._update_feature_set_spec(db_tagged_object, tagged_object_dict)
+
+        self._upsert(session, [db_tagged_object])
+        self.tag_objects_v2(session, [db_tagged_object], project, tag)
+
+        return uid
+
+    def _re_tag_existing_object(
+        self, session, cls, project, name, tag, uid, obj_name_attribute: str = "name"
+    ):
+        if cls == ArtifactV2:
+            _, _, existing_object = self._get_existing_artifact(
+                session, project, name, None, uid
+            )
```

Comment:
you should get the `existing_object` externally and if not given, use _get_record_by_name_tag_and_uid
you can create a wrapper method just for artifacts

Reply:
but the external existing artifact can also be None.